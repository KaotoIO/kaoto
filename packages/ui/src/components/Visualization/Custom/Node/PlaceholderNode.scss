@use '../custom';

.placeholder-node {
  @include custom.highligth {
    &__container {
      @include custom.container;

      display: flex;
      align-items: center;
      justify-content: center;

      &__image {
        border: 2px dashed var(--custom-node-BorderColor);
        border-radius: var(--custom-node-BorderRadius);
        background-color: var(--custom-node-BackgroundColor);
        padding: var(--pf-v5-global--spacer--md);
        color: var(--pf-topology__edge--Stroke);
      }

      &:hover &__image {
        border-color: var(--custom-node-hover-BorderColor);
        box-shadow: var(--custom-node-Shadow);
      }
    }

    &__label {
      @include custom.label {
        background-color: var(--custom-node-BackgroundColor);
        padding: 0 calc(var(--pf-v5-global--spacer--xs) * 0.75);
        border-radius: calc(var(--custom-node-BorderRadius) * 0.4);
        font-style: italic;
      }
    }
  }
}
