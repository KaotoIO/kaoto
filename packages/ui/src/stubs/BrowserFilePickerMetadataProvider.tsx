import { ChangeEvent, createRef, FunctionComponent, PropsWithChildren, useCallback, useRef } from 'react';
import {
  SCHEMA_FILE_ACCEPT_PATTERN,
  SCHEMA_FILE_ACCEPT_PATTERN_SOURCE_BODY,
  SCHEMA_FILE_NAME_PATTERN,
  SCHEMA_FILE_NAME_PATTERN_SOURCE_BODY,
} from '../models/datamapper';
import { IMetadataApi, MetadataContext } from '../providers';
import { readFileAsString } from './read-file-as-string';

export const BrowserFilePickerMetadataProvider: FunctionComponent<PropsWithChildren> = (props) => {
  const fileInputRef = createRef<HTMLInputElement>();
  const fileSelectionRef = useRef<{
    resolve: (files: Record<string, string>) => void;
    reject: (error: unknown) => unknown;
  }>();
  const fileContentsRef = useRef<Record<string, string>>();

  const askUserForFileSelection = useCallback(
    (
      include: string,
      _exclude?: string,
      _options?: Record<string, unknown>,
    ): Promise<string[] | string | undefined> => {
      if (!fileInputRef.current) return Promise.resolve(undefined);

      if (include === SCHEMA_FILE_NAME_PATTERN_SOURCE_BODY) {
        fileInputRef.current.accept = SCHEMA_FILE_ACCEPT_PATTERN_SOURCE_BODY;
      } else if (include === SCHEMA_FILE_NAME_PATTERN) {
        fileInputRef.current.accept = SCHEMA_FILE_ACCEPT_PATTERN;
      }

      fileInputRef.current.click();
      return new Promise<Record<string, string>>((resolve, reject) => {
        fileSelectionRef.current = { resolve, reject };
      }).then((files) => {
        fileContentsRef.current = files;
        return Object.keys(files);
      });
    },
    [fileInputRef],
  );

  const onImport = useCallback(async (event: ChangeEvent<HTMLInputElement>) => {
    const schemaFiles = event.target.files;
    if (!schemaFiles) return;
    const fileContents: Record<string, string> = {};
    const fileContentPromises: Promise<string>[] = [];
    Array.from(schemaFiles).forEach((f) => {
      const promise = readFileAsString(f).then((content) => (fileContents[f.name] = content));
      fileContentPromises.push(promise);
    });
    await Promise.allSettled(fileContentPromises);

    fileSelectionRef.current?.resolve(fileContents);
    fileSelectionRef.current = undefined;
    event.target.value = '';
  }, []);

  const getResourceContent = useCallback((path: string) => {
    return Promise.resolve(fileContentsRef.current && fileContentsRef.current[path]);
  }, []);

  const metadataApi: IMetadataApi = {
    askUserForFileSelection: askUserForFileSelection,
    getResourceContent: getResourceContent,
    shouldSaveSchema: true,
    getMetadata: () => Promise.resolve(undefined),
    setMetadata: () => Promise.resolve(),
    deleteResource: () => Promise.resolve(true),
    saveResourceContent: () => Promise.resolve(),
    getSuggestions: () => Promise.resolve([]),
    onStepUpdated: () => Promise.resolve(),
  };

  return (
    <MetadataContext.Provider value={metadataApi}>
      {props.children}
      <input
        type="file"
        style={{ display: 'none' }}
        data-testid={`attach-schema-file-input`}
        onChange={onImport}
        ref={fileInputRef}
      />
    </MetadataContext.Provider>
  );
};
