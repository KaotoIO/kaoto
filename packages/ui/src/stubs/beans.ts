export const beansYaml = `
- beans:
  - name: myBean
    type: io.kaoto.MyBean
    properties:
      p1: p1v
      p2: p2v
  - name: myBean2
    type: io.kaoto.MyBean
    properties:
      p1:
        p1s1: p1s1v
`

export const beansJson = {
  "beans": [
    {
      "name": "myBean",
      "type": "io.kaoto.MyBean",
      "properties": {
        "p1": "p1v",
        "p2": "p2v"
      }
    },
    {
      "name": "myBean2",
      "type": "io.kaoto.MyBean",
      "properties": {
        "p1": {
          "p1s1": "p1s1v"
        }
      }
    }
  ]
}

