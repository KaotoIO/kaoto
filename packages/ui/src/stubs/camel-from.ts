import { FromDefinition } from '@kaoto-next/camel-catalog/types';
import { CamelRouteVisualEntity } from '../models/visualization/flows';

/**
 * This is a stub Camel From in YAML format.
 * It is used to test the Canvas component.
 */
export const camelFromYaml = `
- from:
    uri: timer:tutorial
    steps:
    - to:
        uri: direct:my-route
`;

/**
 * This is a stub Camel From in JSON format.
 * It is used to test the Canvas component.
 */
export const camelFromJson: { from: FromDefinition } = {
  from: {
    uri: 'timer:tutorial',
    steps: [
      {
        to: {
          uri: 'direct:my-route',
        },
      },
    ],
  },
};

export const camelFrom = new CamelRouteVisualEntity({ from: camelFromJson.from });
