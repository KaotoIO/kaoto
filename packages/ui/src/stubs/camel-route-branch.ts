import { parse } from 'yaml';

export const camelRouteBranch = parse(`
- route:
    id: route-2768
    from:
      id: from-4014
      uri: timer:template
      parameters:
        period: "1000"
      steps:
        - choice:
            id: choice-3431
            otherwise:
              id: otherwise-3653
              steps:
                - log:
                    id: log-6808
                    message: \${body}
            when:
              - id: when-4112
                steps:
                  - setHeader:
                      id: setHeader-6078
                      expression:
                        simple: {}
                expression:
                  simple:
                    expression: \${header.foo} == 1
        - to:
            id: to-3757
            uri: sql
            parameters: {}
`);
