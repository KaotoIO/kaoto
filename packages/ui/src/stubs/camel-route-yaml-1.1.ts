export const camelRouteYaml_1_1_original = `
- route:
    id: route-3376
    from:
      id: from-3505
      uri: rest:post:/newCustomer
      parameters:
        host: localhost
      steps: []
`;

export const camelRouteYaml_1_1_updated = `
- route:
    id: route-3376
    from:
      id: from-3505
      uri: rest:post:/newCustomer
      parameters:
        bindingMode: "off"
        host: localhost
      steps: []
`;
