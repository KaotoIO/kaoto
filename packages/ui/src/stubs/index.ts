export * from './camel-route';
export * from './kamelet-binding-route';
export * from './kamelet-route';
export * from './pipe';
