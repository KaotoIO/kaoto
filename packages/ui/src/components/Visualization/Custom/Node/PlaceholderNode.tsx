import { Icon } from '@patternfly/react-core';
import { PlusCircleIcon } from '@patternfly/react-icons';
import type { DefaultNode, ElementModel, GraphElement, Node } from '@patternfly/react-topology';
import {
  AnchorEnd,
  DEFAULT_LAYER,
  Layer,
  Rect,
  isNode,
  observer,
  useAnchor,
  withDndDrop,
} from '@patternfly/react-topology';
import { FunctionComponent, useContext, useRef } from 'react';
import { IVisualizationNode } from '../../../../models';
import { SettingsContext } from '../../../../providers';
import { CanvasDefaults } from '../../Canvas/canvas.defaults';
import { CanvasNode } from '../../Canvas/canvas.models';
import { useReplaceStep } from '../hooks/replace-step.hook';
import { TargetAnchor } from '../target-anchor';
import './PlaceholderNode.scss';
import clsx from 'clsx';
import { placeholderNodeDropTargetSpec } from '../customComponentUtils';

type DefaultNodeProps = Parameters<typeof DefaultNode>[0];
interface PlaceholderNodeInnerProps extends DefaultNodeProps {
  element: GraphElement<ElementModel, CanvasNode['data']>;
}

const PlaceholderNodeInner: FunctionComponent<PlaceholderNodeInnerProps> = observer(
  ({ element, dndDropRef, hover, canDrop }) => {
    if (!isNode(element)) {
      throw new Error('PlaceholderNode must be used only on Node elements');
    }
    const vizNode: IVisualizationNode | undefined = element.getData()?.vizNode;
    const settingsAdapter = useContext(SettingsContext);
    const label = vizNode?.getNodeLabel(settingsAdapter.getSettings().nodeLabel);
    const updatedLabel = label === 'placeholder' ? 'Add step' : label;
    const tooltipContent = 'Click to add a step';
    const boxRef = useRef<Rect>(element.getBounds());
    const labelX = (boxRef.current.width - CanvasDefaults.DEFAULT_LABEL_WIDTH) / 2;

    useAnchor((element: Node) => {
      return new TargetAnchor(element);
    }, AnchorEnd.both);

    if (!vizNode) {
      return null;
    }
    const { onReplaceNode } = useReplaceStep(vizNode);

    return (
      <Layer id={DEFAULT_LAYER}>
        <g
          className="placeholder-node"
          data-testid={`placeholder-node__${vizNode.id}`}
          data-nodelabel={label}
          onClick={onReplaceNode}
        >
          <foreignObject
            data-nodelabel={label}
            width={boxRef.current.width}
            height={boxRef.current.height}
            ref={dndDropRef}
          >
            <div
              className={clsx('placeholder-node__container', {
                'placeholder-node__container__dropTarget': canDrop && hover,
              })}
            >
              <div title={tooltipContent} className="placeholder-node__container__image">
                <Icon size="lg">
                  <PlusCircleIcon />
                </Icon>
              </div>
            </div>
          </foreignObject>

          <foreignObject
            x={labelX}
            y={boxRef.current.height - 1}
            width={CanvasDefaults.DEFAULT_LABEL_WIDTH}
            height={CanvasDefaults.DEFAULT_LABEL_HEIGHT}
            className="placeholder-node__label"
          >
            <div className="placeholder-node__label__text">
              <span title={updatedLabel}>{updatedLabel}</span>
            </div>
          </foreignObject>
        </g>
      </Layer>
    );
  },
);

export const PlaceholderNode = withDndDrop(placeholderNodeDropTargetSpec)(PlaceholderNodeInner);
