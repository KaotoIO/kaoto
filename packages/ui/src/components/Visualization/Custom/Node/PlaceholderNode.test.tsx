import { BaseEdge, BaseGraph, BaseNode, ElementContext, VisualizationProvider } from '@patternfly/react-topology';
import { act, render } from '@testing-library/react';
import { TestProvidersWrapper } from '../../../../stubs';
import { ControllerService } from '../../Canvas/controller.service';
import { PlaceholderNodeObserver } from './PlaceholderNode';

describe('PlaceholderNode', () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should throw an error if not used on Node elements', () => {
    jest.spyOn(console, 'error').mockImplementation(() => {});
    const edgeElement = new BaseEdge();

    expect(() => {
      act(() => {
        render(<PlaceholderNodeObserver element={edgeElement} />);
      });
    }).toThrow('PlaceholderNode must be used only on Node elements');
  });

  it('should render without error', () => {
    const parentElement = new BaseGraph();
    const element = new BaseNode();
    const controller = ControllerService.createController();
    parentElement.setController(controller);
    element.setController(controller);
    element.setParent(parentElement);

    const { Provider } = TestProvidersWrapper();

    const wrapper = render(
      <Provider>
        <VisualizationProvider controller={controller}>
          <ElementContext.Provider value={element}>
            <PlaceholderNodeObserver element={element} />
          </ElementContext.Provider>
        </VisualizationProvider>
      </Provider>,
    );

    expect(wrapper.asFragment()).toMatchSnapshot();
  });
});
