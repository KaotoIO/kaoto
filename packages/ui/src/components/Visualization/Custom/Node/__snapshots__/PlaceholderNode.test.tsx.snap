// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaceholderNode should render without error 1`] = `<DocumentFragment />`;
